basic_pipeline/09_Mixer.md

```diff
@@ -53,186 +61,149 @@ The logic we're going to implement can be described in the following three steps
 ```
 
-What we do is that we are simply putting the incoming `buffer` into the `Track` structure for the given pad. Note, that we have to be sure that we are not losing any information which is in the `Track`'s buffer before the update. In case there is a buffer on a given `Track`, it has to be processed before another buffer comes. Why can we be sure of that in our implementation? That's before we precisely steer the flow of our program and ask for the next buffer after we empty the buffer hold in the state of the element.
+In this callback we update the mixer's state by assigning the incoming buffer to its track. We can be sure no overwriting of an existing buffer happens, which will become more apparent as we delve further into the logic's implementation. 
+Once the state is updated we gather all buffers that can be sent (might be none) in `get_output_buffers_actions/1` and return the coresponding `buffer` actions. In case any demands should be sent afterwards we also tell the output pad to redemand.
 
-**_`lib/elements/Mixer.ex`_**
+**_`lib/elements/mixer.ex`_**
 
 ```elixir
 defmodule Basic.Elements.Mixer do
  ...
- @impl true
- def handle_end_of_stream(pad, _context, state) do
-  tracks =
-    Map.update!(state.tracks, pad, fn track ->
-      %Track{track | status: :finished}
-    end)
-
-  state = %{state | tracks: tracks}
-  { {:ok, [{:redemand, :output}]}, state}
- end
+  @impl true
+  def handle_end_of_stream(pad, _context, state) do
+    tracks = Map.update!(state.tracks, pad, &%Track{&1 | status: :finished})
+    {tracks, buffer_actions} = get_output_buffers_actions(tracks)
+    state = %{state | tracks: tracks}
+
+    if Enum.all?(tracks, fn {track_id, track} ->
+         track.status == :finished and not has_buffer?({track_id, track})
+       end) do
+      {buffer_actions ++ [end_of_stream: :output], state}
+    else
+      {buffer_actions ++ [redemand: :output], state}
+    end
+  end
  ...
 end
 ```
 
-What we did here was similar to the logic defined in the `handle_process/4` - we have just updated the state of the track (in that case - by setting its status as `:finished`) and then we called the `handle_demand/5` callback using the `:redemand` actions. The `handle_demand/5` will take care of the fact that the track state has changed.
-There is nothing left to do apart from defining the `handle_demand/5` itself!
+What we did here was similar to the logic defined in `handle_buffer/4` - we have just updated the state of the track (in that case - by setting its status as `:finished`), gather the buffers and send them. The important difference is that in case all inputs have closed, we should forward an `end_of_stream` action instead of a `redemand`, signaling the mixer has finished its processing. 
+The `has_buffer?/1` function is a private utility that will show up in a few more places in our element's definition and is nothing more than a simple check `track.buffer != nil`.
+Let's now implement the `handle_demand/5` callback:
```

Comment:
Perhaps it would be better to show and describe private functions used in the `handle_buffer/4` implementation, and then continue with `handle_demand/5`, WDYT?